presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/TDigestAndPercentileStateFactory.java

```diff
@@ -0,0 +1,156 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.state;
+
+import com.facebook.presto.array.ObjectBigArray;
+import com.facebook.presto.spi.function.AccumulatorStateFactory;
+import com.facebook.presto.tdigest.TDigest;
+import io.airlift.slice.SizeOf;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class TDigestAndPercentileStateFactory
+        implements AccumulatorStateFactory<TDigestAndPercentileState>
+{
+    @Override
+    public TDigestAndPercentileState createSingleState()
+    {
+        return new SingleTDigestAndPercentileState();
+    }
+
+    @Override
+    public Class<? extends TDigestAndPercentileState> getSingleStateClass()
+    {
+        return SingleTDigestAndPercentileState.class;
+    }
+
+    @Override
+    public TDigestAndPercentileState createGroupedState()
+    {
+        return new GroupedTDigestAndPercentileState();
+    }
+
+    @Override
+    public Class<? extends TDigestAndPercentileState> getGroupedStateClass()
+    {
+        return GroupedTDigestAndPercentileState.class;
+    }
+
+    public static class GroupedTDigestAndPercentileState
+            extends AbstractGroupedAccumulatorState
+            implements TDigestAndPercentileState
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedTDigestAndPercentileState.class).instanceSize();
+        private final ObjectBigArray<TDigest> digests = new ObjectBigArray<>();
+        private final ObjectBigArray<List<Double>> percentilesArray = new ObjectBigArray<>();
+        private long size;
+
+        @Override
+        public void ensureCapacity(long size)
+        {
+            digests.ensureCapacity(size);
+            percentilesArray.ensureCapacity(size);
+        }
+
+        @Override
+        public TDigest getDigest()
+        {
+            return digests.get(getGroupId());
+        }
+
+        @Override
+        public void setDigest(TDigest digest)
+        {
+            digests.set(getGroupId(), requireNonNull(digest, "digest is null"));
+        }
+
+        @Override
+        public List<Double> getPercentiles()
+        {
+            return percentilesArray.get(getGroupId());
+        }
+
+        @Override
+        public void setPercentiles(List<Double> percentiles)
+        {
+            percentilesArray.set(getGroupId(), requireNonNull(percentiles, "percentiles is null"));
```

Comment:
This will create a List => Double for every group, which could be in the billions.  Most people use this function with static quantiles.  Can we change it so we have just one List => Double for the quantiles and validate that they're static in the function?